### PR TITLE
stdlib_hash: waterhash algorithm

### PR DESCRIPTION
**Pull request by awvwgk**
Implementation of [waterhash](https://github.com/tommyettinger/waterhash) algorithm. Returns a 32 bit hash using a 64 bit seed.

Fpm project available at [awvwgk/waterhash](https://github.com/awvwgk/waterhash) for review.

### CODE REVIEW

**Review by hsnyder — APPROVED** [review]
Looks good to me! Though, should we perhaps add check_eq to our stdlib_error module?

**Review by 14NGiestas — APPROVED** [review]

**Comment by awvwgk** [comment]
Thanks for reviewing @hsnyder and @14NGiestas.

@fortran-lang/stdlib We have to approvals for this patch, how do we want to move forward here?

**Comment by milancurcic** [comment]
Can the reviewers please approve and then I suggest merging when ready?

**Comment by jvdp1** [comment]
I have some suggestions for this PR. However, I have not much time currently for focussing on it now. Probably next week. However, if needed, I am ok to merge as is, and review it later.

**Comment by wclodius2** [comment]
FWIW I have my own implementation of the water hash. It differs from yours in that:

1. Your argument p is named key.
2. It  attempts to make the result endian independent.
3. It  provides wrappers for arrays of INT8, INT16, INT32,  and INT64.
4. It attempts to deal with default characters that need not be 8  bits.
5. It  provides more detailed  attributions.

**Comment by wclodius2** [comment]
See also my latest posting to issue #262

**Review by 14NGiestas — APPROVED** [review]
This patch is well structured and as far as I tested it worked well. I like how it allows further enhancements and new hashes. The usage is also pretty straightforward.

**Comment by milancurcic** [comment]
@wclodius2 if this PR went forward, could this implementation be improved with the differences that you list in a separate PR?

**Comment by wclodius2** [comment]
Yes it could be improved. To see what my proposal looks like see `stdlib_32_bit_water_hashes.fypp` at the [`hash_function_api`](https://github.com/wclodius2/hash_function_api) directory.

**Comment by jvdp1** [comment]
Thank you @wclodius2 for your update.
Based on your Github directory and included specs, I would suggest that @awvwgk closes this PR (or move to a draft) and that @wclodius2 opens a new PR with his implemented hash functions. I also like that @wclodius2 kept all the associated licenses (compatible with MIT AFAIK).
@milancurcic @awvwgk what do you think? How should we move with this topic?

**Comment by awvwgk** [comment]
Sure, I'll close this PR.

**Pull request closed without merging**
